zookeeper-server/src/test/java/org/apache/zookeeper/server/RateLoggerTest.java

```diff
@@ -0,0 +1,163 @@
+
+    /** Verify that a single message with value is logged. */
+    @Test
+    public void singleMessageWithValueExplicitFlush() {
+        RateLogger rateLogger = new RateLogger(LOGGER, 5 * 1000L);
+        rateLogger.rateLimitLog("test message", "testValue");
+        rateLogger.flush();
+        assertThat(testAppender.messages,
+                contains("Message: test message, Value: testValue"));
+    }
+
+    /** Verify that a repeated message and value is not logged until flushed. */
+    @Test
+    public void recurringMessageWithValueExplicitFlush() {
+        RateLogger rateLogger = new RateLogger(LOGGER, 5 * 1000L);
+        rateLogger.rateLimitLog("test message", "value-one");
+        rateLogger.rateLimitLog("test message", "value-two");
+        assertThat(testAppender.messages,
+                contains("Message: test message, Value: value-one"));
+        // flush logs second message
+        rateLogger.flush();
+        assertThat(testAppender.messages,
+                contains("Message: test message, Value: value-one",
+                        "Message: test message, Last value: value-two"));
+    }
+
+    /** Verify that a message change writes the previous message. */
+    @Test
+    public void messageChangeImplicitFlush() {
+        RateLogger rateLogger = new RateLogger(LOGGER, 5 * 1000L);
+        rateLogger.rateLimitLog("test message one");
+        rateLogger.rateLimitLog("test message one");
+        rateLogger.rateLimitLog("test message two");
+        assertThat(testAppender.messages,
+                contains("Message: test message one",
+                         "Message: test message one",
+                         "Message: test message two"));
+    }
+
+    /** Verify that a count is included for more than one message and value. */
+    @Test
+    public void recurringMessageWithValueImplicitFlush() {
+        RateLogger rateLogger = new RateLogger(LOGGER, 5 * 1000L);
+        rateLogger.rateLimitLog("test message one", "value-one");
+        rateLogger.rateLimitLog("test message one", "value-two");
+        rateLogger.rateLimitLog("test message one", "value-three");
+        rateLogger.flush();
+        assertThat(testAppender.messages,
+                contains("Message: test message one, Value: value-one",
+                        "[2 times] Message: test message one, Last value: value-three"));
+    }
+
+    /** Verify the a message written after the rate interval logs the previous message. */
+    @Test
+    public void recurringMessageTimedImplicitFlush() throws InterruptedException {
+        long rateLogInterval = 500L;
+        RateLogger rateLogger = new RateLogger(LOGGER, rateLogInterval);
+        rateLogger.rateLimitLog("test message one", "value-one");
+        rateLogger.rateLimitLog("test message one", "value-two");
+        rateLogger.rateLimitLog("test message one", "value-three");
+        Thread.sleep(2 * rateLogInterval);
+        // same message after rate interval triggers implicit flush
```

Comment:
this test may be flaky, you can look at other example how to mock timing (e.g.: `CreateContainerTest#testMaxNeverUsedInterval`)

Reply:
Updated RateLogger to get current time from a method with package access that is overridden during unit testing.  This allows the time to be explicitly set similar to the way you have suggested.